Fix login failure with long OpenId tokens

israel-hdez opened the pull request:
In the authorization code flow, tokens returned by IdP can be long to the point that the session cookie also becomes long. Although RFCs don't set a maximum cookie size (see https://tools.ietf.org/html/rfc2965), in practice, major browsers use the RFC minimum limit as it's supported maximum - which is 4K. This means that longs tokens can cause the session token to reach this limit and be discarded by browsers.

This is overcoming these limitations by splitting session data in chunk no longer than 4K and setting multiple cookies, if needed.

The multi-cookie approach is applied only to the "authorization code" flow of OpenID. The "implicit flow" is left unchanged because IDPs tend to omit group membership data, which is the data that users have reported to generate larger tokens.

Fixes #3874

jmazzitelli commented on handlers/authentication.go:
```diff
@@ -924,15 +949,46 @@ func OpenIdCodeFlowHandler(w http.ResponseWriter, r *http.Request) bool {
+		var cookieName string
+		if i == 0 {
+			// Set a cookie with the regular cookie name with the first chunk of session data.
+			// This is for backwards compatibility
```
It's also nice that in most use-cases, you still only ever have one cookie as opposed to also needing a `-chunks` cookie with a value of 1.

jmazzitelli replied:
or do you always still have a `-chunks` cookie? I think this implementation does. Oh, well, never mind :)

israel-hdez replied:
It's always there!
But it would be a simple condition to add around these lines: https://github.com/kiali/kiali/pull/3898/files#diff-3dd5896ffb3bb3685544e6a40338180ed95a5864b929b908fd1a70ec7628b251R983-R991

I can put that condition if you prefer not having the `-chunks` with value "1" :-)

jmazzitelli replied:
You can leave it alone. "No biggie"

jmazzitelli reviewed (approved):
I looked at the code and it all looks correct to me. I don't have an openid setup so I didn't run this.

israel-hdez commented:
@somejfn I wonder if you could try this image and check that the issue with Keycloak is gone: quay.io/edgarhz/kiali:rc-20210416-1131

@kulbhusanazad perhaps can you also try that image ^^? It includes PR #3838 which are the changes to support Azure. In the operator's KialiCR you should be able to configure these new options:

```yaml
spec:
  auth:
    openid:
      api_token: "access_token"
      additional_request_params:
        resource: {your_resource_id}
```

...and all your other configs should be the same.

---

If you can confirm this works, this can go in today's Kiali release. But try ASAP.

israel-hdez commented:
@jmazzitelli Ouch! I forgot the code to properly do "logout".
I added the code to a new commit. So, I think this deserves a follow-up review from you.

Also, I added the conditional to omit the "-chunks" cookie if there is only one chunk to keep number of cookies to a minimum ;-)

jmazzitelli commented on business/openid_auth.go:
```diff
@@ -238,7 +245,49 @@ func GetOpenIdAesSession(r *http.Request) (*config.IanaClaims, error) {
+
+		// It's known that major browsers have a limit of 180 cookies per domain.
+		if numChunks <= 0 || numChunks > 180 {
+			return nil, fmt.Errorf("number of session cookies is %d, but limit is 180", numChunks)
```
```suggestion
			return nil, fmt.Errorf("number of session cookies is %d, but limit is 1 through 180", numChunks)
```

jmazzitelli replied:
```suggestion
			return nil, fmt.Errorf("number of session cookies is %d, but limit is 1 through 180", numChunks)
```

jmazzitelli reviewed (approved):
I'll approve with that minor comment/suggestion, in case you want to do that.
Again, I didn't run, but I see what's going on and the code looks ok.

My only question is that in the code I see we continue even if a -aes cookie is missing (`log.Warningf("Apparently, there is no AES session: %s ", aes.Error())` but yet the deleteTokenCookies function always drops it. I think this is OK (you are just telling the client to drop a cookie that might not exist - you would think that would be a no-op) but I bring it up just in case.

israel-hdez commented:
> My only question is that in the code I see we continue even if a -aes cookie is missing (`log.Warningf("Apparently, there is no AES session: %s ", aes.Error())` but yet the deleteTokenCookies function always drops it. I think this is OK (you are just telling the client to drop a cookie that might not exist - you would think that would be a no-op) but I bring it up just in case.

@jmazzitelli I'm not sure if I understand your question, but I'll answer this:

The `AuthenticationInfo` func where this `log.Warning` located, is expected to be called by the front-end to check if there is an active session. If there is a session, the front-end expects some user details (e.g. his display name) and it will show the dashboard rather than the login page. Then, if there is no session, the back-end sends empty user info and the front-end requests a login. So, we want to gracefully fail here.

In the `deleteTokenCookies`, yes it's a no-op. Indeed, it checks for cookie presence to send the deletion. See: https://github.com/kiali/kiali/pull/3898/files#diff-3dd5896ffb3bb3685544e6a40338180ed95a5864b929b908fd1a70ec7628b251R1028
